src/backends/gitlab/API.js

```diff
@@ -0,0 +1,196 @@
+import LocalForage from "Lib/LocalForage";
+import { Base64 } from "js-base64";
+import { fromJS, List, Map } from "immutable";
+import { cond, flow, isString, partial, partialRight, omit, set, update } from "lodash";
+import unsentRequest from "Lib/unsentRequest";
+import { then } from "Lib/promiseHelper";
+import AssetProxy from "ValueObjects/AssetProxy";
+import { APIError } from "ValueObjects/errors";
+import Cursor from "ValueObjects/Cursor"
+
+export default class API {
+  constructor(config) {
+    this.api_root = config.api_root || "https://gitlab.com/api/v4";
+    this.token = config.token || false;
+    this.branch = config.branch || "master";
+    this.repo = config.repo || "";
+    this.repoURL = `/projects/${ encodeURIComponent(this.repo) }`;
+  }
+
+  withAuthorizationHeaders = req =>
+    unsentRequest.withHeaders(this.token ? { Authorization: `Bearer ${ this.token }` } : {}, req);
+
+  buildRequest = req => flow([
+    unsentRequest.withRoot(this.api_root),
+    this.withAuthorizationHeaders,
+    unsentRequest.withTimestamp,
+  ])(req);
+
+  request = async req => flow([this.buildRequest, unsentRequest.performRequest])(req);
+  requestURL = url => flow([unsentRequest.fromURL, this.request])(url);
+  responseToJSON = res => res.json()
+  responseToText = res => res.text()
+  requestJSON = req => this.request(req).then(this.responseToJSON);
+  requestText = req => this.request(req).then(this.responseToText);
+  requestJSONFromURL = url => this.requestURL(url).then(this.responseToJSON);
+  requestTextFromURL = url => this.requestURL(url).then(this.responseToText);
+
+  user = () => this.requestJSONFromURL("/user");
+
+  WRITE_ACCESS = 30;
+  hasWriteAccess = user => this.requestJSONFromURL(this.repoURL).then(({ permissions }) => {
+    const { project_access, group_access } = permissions;
+    if (project_access && (project_access.access_level >= this.WRITE_ACCESS)) {
+      return true;
+    }
+    if (group_access && (group_access.access_level >= this.WRITE_ACCESS)) {
+      return true;
+    }
+    return false;
+  });
+
+  readFile = async (path, sha, ref=this.branch) => {
+    const cachedFile = sha ? await LocalForage.getItem(`gl.${ sha }`) : null;
```

Comment:
Just wondering whether we should really be differentiating the cache between GitLab and GitHub here (`gh.sha` and `gl.sha`). If the SHA is the same, you've got the same file, regardless of which backend you are using.

Reply:
I agree, but would add two thoughts:

Reply:
Agreed, but two thoughts:

1. If we combine, we should use something besides `gh` or `gl`, maybe just `refs` or `sha`.
2. It might not be a bad idea to leave it as `gl` for right now, then switch at 2.0. Helps make it clear which backend set the sha for troubleshooting, and less work to get this merged.